src/foam/dao/FixedSizeDAO.js

```diff
@@ -46,24 +62,57 @@ foam.CLASS({
         }
       ],
       javaCode: `
+Integer insertAt;
 foam.core.FObject delegatedObject = getDelegate().put_(x, obj);
-getFixedSizeArray().add(delegatedObject);
+getLock().lock();
+try {
+  insertAt = getNextIndex();  
+  if ( insertAt == getInternalArraySize() ) {
+    insertAt = 0;
+  }
+  setNextIndex( insertAt + 1 );
+} finally {
+  getLock().unlock();
+}
+getFixedSizeArray()[insertAt] = delegatedObject;
 return delegatedObject;
   `
     },
     {
       name: 'select_',
       javaReturns: 'foam.dao.Sink',
       javaCode: `
+if ( sink == null ) sink = new ArraySink();
 sink = prepareSink(sink);
 Sink decorated = decorateSink_(sink, skip, limit, order, predicate);
-Object[] arrayObject = getFixedSizeArray().toArray();
-for ( int i = 0; i < arrayObject.length; i++ ) {
-  decorated.put( arrayObject[i], null );
+
+Integer backCounter;
+
+if ( getNextIndex() == 0 ) {
+  backCounter = ( getInternalArraySize() - 1 ); 
+} else {
+  backCounter = ( getNextIndex() - 1 );
```

Comment:
Check that backCounter is not -1 after this.

Reply:
This section to prevent precisely that. 
It could only be -1 if nextIndex was zero, which I'm checking for. 
I don't think it should be necessary to do another explicit check...

Reply:
This is outside the lock so it could be reset between the two calls.

Reply:
Makes sense, made the change.